lib/puppet/defaults.rb

```diff
@@ -727,13 +727,34 @@ def self.initialize_default_settings!(settings)
         unless [0, 'unlimited', Float::INFINITY].include?(val)
           Puppet.deprecation_warning(<<-WARNING)
 Fine grained control of environment timeouts is deprecated,
-please use `0` or `unlimited` to control default caching behavior
-and the environment-cache endpoint in Puppet Server's administrative
-API to expire the cache as needed
+please use `environment_ttl` instead.
           WARNING
         end
       end
     },
+    :environment_ttl => {
+      :default    => "unlimited",
```

Comment:
This needs to default to 0 to give a consistent experience. But if both `environment_ttl` and `environment_timeout` default to 0, then users have to set both to non-zero values (in order to cache environments). Or we need to detect if either setting is explicitly set (as opposed to inheriting the default value).

I think the cleanest approach would be to always use `environment_ttl` if it's explicitly set. Otherwise use `environment_timeout` as before. Setting `environment_ttl` to `0` or `unlimited` will work the same as `environment_timeout`. Setting `environment_ttl` to a value between those will yield the desired MRU behavior. Feature flagging the behavior based on `environment_ttl` also makes upgrades to 7 easier, since the behavior of `environment_ttl` is not dependent on `environment_timeout` which will be removed in puppet 7.